File: app/blacklist.py
"""E621 tag blacklist parsing and cluster evaluation logic."""

import base64
from fnmatch import fnmatch
import re
from typing import Any
from urllib.parse import unquote

RATING_MAP: dict[str, str] = {
    "s": "s",
    "safe": "s",
    "q": "q",
    "questionable": "q",
    "e": "e",
    "explicit": "e",
}

RATING_ORDER: dict[str, int] = {"e": 3, "q": 2, "s": 1}


class TermMatcher:
    """Matches a single tag term, wildcard pattern, or rating specifier."""

    def __init__(self, raw_term: str) -> None:
        self.raw_term: str = raw_term.lower().strip()
        self.is_rating: bool = self.raw_term.startswith("rating:")
        self.allowed_ratings: set[str] = set()
        self.has_wildcard: bool = False

        if self.is_rating:
            rating_val = self.raw_term.split(":", 1)[1]
            parts = rating_val.split(",")
            self.allowed_ratings = {
                RATING_MAP.get(p.strip(), p.strip()) for p in parts if p.strip()
            }
        else:
            self.has_wildcard = "*" in self.raw_term

    def matches(self, tags: set[str], rating: str) -> bool:
        if self.is_rating:
            return rating in self.allowed_ratings
        if self.has_wildcard:
            return any(fnmatch(t, self.raw_term) for t in tags)
        return self.raw_term in tags


class BlacklistRule:
    """Represents a single parsed line/rule from an e621 blacklist."""

    def __init__(self, line: str) -> None:
        self.raw_line: str = line.strip()
        self.positive_terms: list[TermMatcher] = []
        self.negative_terms: list[TermMatcher] = []
        self.or_groups: list[list[TermMatcher]] = []
        self._parse(line)

    def _parse(self, line: str) -> None:
        if "#" in line:
            line = line.split("#", 1)[0]
        line = line.strip()
        if not line:
            return

        def parse_prefix(tok: str) -> tuple[str, str]:
            if tok.startswith("-"):
                return ("-", tok[1:])
            if tok.startswith("~"):
                return ("~", tok[1:])
            return ("", tok)

        tokens: list[str] = re.findall(
            r"-\(\s*[^)]+\s*\)|\(\s*[^)]+\s*\)|[^\s()]+", line
        )
        standalone_or_group: list[TermMatcher] = []

        for token in tokens:
            token = token.strip()
            if not token:
                continue

            if token.startswith("-(") and token.endswith(")"):
                inner = token[2:-1].strip()
                for itok in inner.split():
                    _, clean = parse_prefix(itok)
                    self.negative_terms.append(TermMatcher(clean))

            elif token.startswith("(") and token.endswith(")"):
                inner = token[1:-1].strip()
                group: list[TermMatcher] = []
                for itok in inner.split():
                    _, clean = parse_prefix(itok)
                    group.append(TermMatcher(clean))
                if group:
                    self.or_groups.append(group)

            else:
                prefix, clean = parse_prefix(token)
                if prefix == "-":
                    if standalone_or_group:
                        self.or_groups.append(standalone_or_group)
                        standalone_or_group = []
                    self.negative_terms.append(TermMatcher(clean))

                elif prefix == "~":
                    standalone_or_group.append(TermMatcher(clean))

                else:
                    if standalone_or_group:
                        self.or_groups.append(standalone_or_group)
                        standalone_or_group = []
                    self.positive_terms.append(TermMatcher(clean))

        if standalone_or_group:
            self.or_groups.append(standalone_or_group)

    def is_empty(self) -> bool:
        return not (
            self.positive_terms or self.negative_terms or self.or_groups
        )

    def matches(self, tags: set[str], rating: str) -> bool:
        if self.is_empty():
            return False

        for term in self.positive_terms:
            if not term.matches(tags, rating):
                return False

        for term in self.negative_terms:
            if term.matches(tags, rating):
                return False

        for group in self.or_groups:
            if not any(term.matches(tags, rating) for term in group):
                return False

        return True


class E621BlacklistEvaluator:
    """Evaluates text/tag rules against target post/cluster metadata."""

    def __init__(self, blacklist_text: str) -> None:
        self.rules: list[BlacklistRule] = []
        for line in blacklist_text.splitlines():
            rule = BlacklistRule(line)
            if not rule.is_empty():
                self.rules.append(rule)

    def evaluate(self, tags: set[str], rating: str) -> tuple[bool, str | None]:
        """Returns (is_blacklisted, matching_rule_line)."""
        for rule in self.rules:
            if rule.matches(tags, rating):
                return True, rule.raw_line
        return False, None


def decode_blacklist_header(header_val: str | None) -> str | None:
    """Decodes a URL-encoded or Base64-encoded blacklist header string."""
    if not header_val:
        return None

    # Try Base64 decoding first
    try:
        return base64.b64decode(header_val).decode("utf-8")
    except Exception:
        pass

    # Fall back to standard URL unquoting
    try:
        return unquote(header_val)
    except Exception:
        return header_val


def get_cluster_union_tags_and_rating(
    cluster_posts: list[dict[str, Any]],
) -> tuple[set[str], str]:
    """Extracts union set of tags and canonical highest rating (e > q > s)."""
    union_tags: set[str] = set()
    max_rating_score = 0
    canonical_rating = "s"

    for post in cluster_posts:
        p_rating = (post.get("rating") or "s").lower()
        score = RATING_ORDER.get(p_rating, 1)
        if score > max_rating_score:
            max_rating_score = score
            canonical_rating = p_rating

        tags_dict: dict[str, list[str]] = post.get("tags_json") or {}

        for cat_tags in tags_dict.values():
            for tag in cat_tags:
                union_tags.add(tag.lower())

    return union_tags, canonical_rating

File: static/js/blacklist.js
// @ts-check

import Alpine from 'alpinejs';
import { showToast } from './toasts.js';
import { fetchCurrentUserProfile } from './e621_api.js';
import { getE621User } from './auth.js';

const DEFAULT_BLACKLIST = `# Violence
gore
snuff
rape

# ABDL
young -rating:s
diaper -rating:s

# Fetish
feces
urine
fart_fetish
realistic_feral rating:e

# Controversial
politics`;

/** @type {Record<string, PostRating>} */
const RATING_MAP = {
    s: 's', safe: 's',
    q: 'q', questionable: 'q',
    e: 'e', explicit: 'e',
};

/** @type {Record<PostRating, number>} */
const RATING_ORDER = { s: 1, q: 2, e: 3 };

/**
 * Normalizes e621 ratings into 's', 'q', or 'e'.
 * @param {string} r
 * @returns {PostRating}
 */
export function normalizeRating(r) {
    if (!r) return 's';
    const clean = String(r).toLowerCase().trim();
    return RATING_MAP[clean] || clean;
}

export class TermMatcher {
    /**
     * @param {string} rawTerm
     */
    constructor(rawTerm) {
        this.rawTerm = rawTerm.toLowerCase().trim();
        this.isRating = this.rawTerm.startsWith('rating:');
        /** @type {Set<string>} */
        this.allowedRatings = new Set();
        this.hasWildcard = false;
        /** @type {RegExp|null} */
        this.regex = null;

        if (this.isRating) {
            const ratingVal = this.rawTerm.split(':', 2)[1] || '';
            for (const p of ratingVal.split(',')) {
                const clean = p.trim();
                if (clean) this.allowedRatings.add(normalizeRating(clean));
            }
        } else {
            this.hasWildcard = this.rawTerm.includes('*');
            if (this.hasWildcard) {
                const escaped = this.rawTerm
                    .replace(/[.+^${}()|[\]\\]/g, '\\$&')
                    .replace(/\*/g, '.*');
                this.regex = new RegExp(`^${escaped}$`, 'i');
            }
        }
    }

    /**
     * @param {Set<string>} tagsSet
     * @param {PostRating} rating
     * @returns {boolean}
     */
    matches(tagsSet, rating) {
        if (this.isRating) return this.allowedRatings.has(normalizeRating(rating));
        if (this.hasWildcard && this.regex) {
            for (const tag of tagsSet) {
                if (this.regex.test(tag)) return true;
            }
            return false;
        }
        return tagsSet.has(this.rawTerm);
    }
}

export class BlacklistRule {
    /**
     * @param {string} line
     */
    constructor(line) {
        this.rawLine = line.trim();
        /** @type {TermMatcher[]} */
        this.positiveTerms = [];
        /** @type {TermMatcher[]} */
        this.negativeTerms = [];
        /** @type {TermMatcher[][]} */
        this.orGroups = [];
        this._parse(line);
    }

    /**
     * @private
     * @param {string} line
     */
    _parse(line) {
        if (line.includes('#')) line = line.split('#')[0];
        line = line.trim();
        if (!line) return;

        /**
         * @param {string} tok
         * @returns {[string, string]}
         */
        const parsePrefix = (tok) => {
            if (tok.startsWith('-')) return ['-', tok.slice(1)];
            if (tok.startsWith('~')) return ['~', tok.slice(1)];
            return ['', tok];
        };

        const tokens = line.match(/-\(\s*[^)]+\s*\)|\(\s*[^)]+\s*\)|[^\s()]+/g) || [];
        /** @type {TermMatcher[]} */
        let standaloneOrGroup = [];

        for (let token of tokens) {
            token = token.trim();
            if (!token) continue;

            if (token.startsWith('-(') && token.endsWith(')')) {
                const inner = token.slice(2, -1).trim();
                for (const itok of inner.split(/\s+/)) {
                    const [, clean] = parsePrefix(itok);
                    if (clean) this.negativeTerms.push(new TermMatcher(clean));
                }
            } else if (token.startsWith('(') && token.endsWith(')')) {
                const inner = token.slice(1, -1).trim();
                const group = [];
                for (const itok of inner.split(/\s+/)) {
                    const [, clean] = parsePrefix(itok);
                    if (clean) group.push(new TermMatcher(clean));
                }
                if (group.length > 0) this.orGroups.push(group);
            } else {
                const [prefix, clean] = parsePrefix(token);
                if (!clean) continue;

                if (prefix === '-') {
                    if (standaloneOrGroup.length > 0) {
                        this.orGroups.push(standaloneOrGroup);
                        standaloneOrGroup = [];
                    }
                    this.negativeTerms.push(new TermMatcher(clean));
                } else if (prefix === '~') {
                    standaloneOrGroup.push(new TermMatcher(clean));
                } else {
                    if (standaloneOrGroup.length > 0) {
                        this.orGroups.push(standaloneOrGroup);
                        standaloneOrGroup = [];
                    }
                    this.positiveTerms.push(new TermMatcher(clean));
                }
            }
        }

        if (standaloneOrGroup.length > 0) {
            this.orGroups.push(standaloneOrGroup);
        }
    }

    isEmpty() {
        return (
            this.positiveTerms.length === 0 &&
            this.negativeTerms.length === 0 &&
            this.orGroups.length === 0
        );
    }

    /**
     * @param {Set<string>} tagsSet
     * @param {PostRating} rating
     * @returns {boolean}
     */
    matches(tagsSet, rating) {
        if (this.isEmpty()) return false;
        const normRating = normalizeRating(rating);

        for (const term of this.positiveTerms) {
            if (!term.matches(tagsSet, normRating)) return false;
        }
        for (const term of this.negativeTerms) {
            if (term.matches(tagsSet, normRating)) return false;
        }
        for (const group of this.orGroups) {
            if (!group.some((term) => term.matches(tagsSet, normRating))) return false;
        }
        return true;
    }
}

export class E621BlacklistEvaluator {
    /**
     * @param {string} blacklistText
     */
    constructor(blacklistText) {
        /** @type {BlacklistRule[]} */
        this.rules = [];
        if (blacklistText) {
            for (const line of blacklistText.split('\n')) {
                const rule = new BlacklistRule(line);
                if (!rule.isEmpty()) {
                    this.rules.push(rule);
                }
            }
        }
    }

    /**
     * @param {Set<string>} tagsSet
     * @param {PostRating} rating
     * @returns {{ isBlacklisted: boolean, matchedRule: string|null }}
     */
    evaluate(tagsSet, rating) {
        const normRating = normalizeRating(rating);
        for (const rule of this.rules) {
            if (rule.matches(tagsSet, normRating)) {
                return { isBlacklisted: true, matchedRule: rule.rawLine };
            }
        }
        return { isBlacklisted: false, matchedRule: null };
    }
}

/**
 * Extracts union of all tags and worst-case canonical rating for a cluster of posts.
 * @param {ClusterPost[]} clusterPosts
 * @returns {{ unionTags: Set<string>, canonicalRating: PostRating }}
 */
export function getClusterUnionTagsAndRating(clusterPosts) {
    const unionTags = new Set();
    let maxRatingScore = 0;
    /** @type { PostRating } */
    let canonicalRating = 's';

    if (!clusterPosts) return { unionTags, canonicalRating };

    for (const post of clusterPosts) {
        const pRating = normalizeRating(post.rating);
        const score = RATING_ORDER[pRating] || 1;
        if (score > maxRatingScore) {
            maxRatingScore = score;
            canonicalRating = pRating;
        }

        if (Array.isArray(post.tags)) {
            for (const tag of post.tags) {
                unionTags.add(tag);
            }
        }
    }

    return { unionTags, canonicalRating };
}

/**
 * Evaluates a cluster against a blacklist evaluator and sets metadata directly.
 * @param {Cluster} cluster
 * @param {E621BlacklistEvaluator} evaluator
 */
export function applyBlacklistToCluster(cluster, evaluator) {
    const { unionTags, canonicalRating } = getClusterUnionTagsAndRating(cluster.posts);
    const { isBlacklisted, matchedRule } = evaluator.evaluate(unionTags, canonicalRating);

    cluster.canonical_rating = canonicalRating;
    cluster.is_blacklisted = isBlacklisted;
    cluster.matched_rule = matchedRule;
}

export class BlacklistManager {
    constructor() {
        this.blacklistText = localStorage.getItem('e621_blacklist') || DEFAULT_BLACKLIST;
        this.isImportingBlacklist = false;
        this.showBlacklistModal = false;
    }

    openModal() {
        this.showBlacklistModal = true;
    }

    closeModal() {
        this.showBlacklistModal = false;
    }

    /**
     * Creates an evaluator using current blacklist text
     * @returns {E621BlacklistEvaluator}
     */
    getEvaluator() {
        return new E621BlacklistEvaluator(this.blacklistText || '');
    }

    /**
     * Saves current blacklist text to localStorage and recalculates cluster blacklist status.
     * @param {Batch[]} [batches] Array of batches to re-evaluate
     * @param {boolean} [silent=false]
     */
    async saveBlacklist(batches = [], silent = false) {
        localStorage.setItem('e621_blacklist', this.blacklistText || '');
        this.showBlacklistModal = false;

        const evaluator = this.getEvaluator();

        if (Array.isArray(batches)) {
            for (const batch of batches) {
                if (!batch.clusters) continue;
                for (const cluster of batch.clusters) {
                    applyBlacklistToCluster(cluster, evaluator);
                    cluster.collapsed = cluster.is_resolved || cluster.is_blacklisted;
                }
            }
        }

        if (!silent) {
            showToast('Blacklist updated.', 'success');
        }
    }

    /**
     * Imports blacklisted tags from the logged-in user's e621 account profile.
     * @param {Batch[]} [batches] Optional array of batches to re-evaluate after import
     */
    async importBlacklist(batches = []) {
        if (!getE621User()) {
            showToast('You must be logged in to import your e621 blacklist.', 'warning');
            return;
        }

        this.isImportingBlacklist = true;

        try {
            const userData = await fetchCurrentUserProfile();

            if (userData.blacklisted_tags !== undefined) {
                this.blacklistText = userData.blacklisted_tags;
                await this.saveBlacklist(batches, true);
                showToast('Blacklist imported and saved successfully', 'success');
            } else {
                throw new Error('Could not locate blacklisted tags in profile.');
            }
        } catch (err) {
            console.error('[Blacklist] Import error:', err);
            const message = err instanceof Error ? err.message : String(err);
            showToast(`Import failed: ${message}`, 'error');
        } finally {
            this.isImportingBlacklist = false;
        }
    }
}

// Singleton instance
export const blacklistManager = new BlacklistManager();

/**
 * Helper to get an active E621BlacklistEvaluator anywhere in plain JS modules
 * @returns {E621BlacklistEvaluator}
 */
export function getBlacklistEvaluator() {
    if (window.Alpine?.store('blacklist')) {
        // @ts-expect-error - Alpine's store API returns 'unknown'
        return window.Alpine.store('blacklist').getEvaluator();
    }
    return blacklistManager.getEvaluator();
}

// Register as Alpine Store
document.addEventListener('alpine:init', () => {
    Alpine.store('blacklist', blacklistManager);
});